R/tags.R

```diff
@@ -288,6 +295,100 @@ tagFunction <- function(func) {
+#' change their attributes (e.g., change CSS classes) depending upon the context
+#' in which they're rendered (e.g., use one set of CSS classes in one a page
+#' layout, but a different set in another page layout). In this situation,
+#' [tagRenderHook()] is preferable to [tagFunction()] since the latter is more a
+#' "black box" in the sense that you don't know anything about the tag structure
+#' until it's rendered.
+#'
+#' @param tag A [`tag()`] object.
+#' @param func A function to call when the `tag` is rendered. This function
+#'   should have at least one argument (the `tag`) and return anything that can
+#'   be converted into tags via [as.tags()].
+#' @param add If `TRUE`, the previous render function is called before calling
+#'   this `func`. Otherwise, any previous render function is ignored.
+#' @return A [tag()] object with a `.renderHook` field containing `func`. When
+#'   the return value is _rendered_ (such as with [`as.tags()`]), this function
+#'   will be called just prior to writing of HTML.
+#' @export
+#' @seealso [tagFunction()]
+#' @examples
+#' # Have a place holder div and return a span instead
+#' obj <- div("example", .renderHook = function(x) {
+#'   x$name <- "span"
+#'   x
+#' })
+#' obj$name # "div"
+#' print(obj) # Prints as a `span`
+#'
+#' # Add a class to the tag
+#' # Should print a `span` with class `"extra"`
+#' spanExtra <- tagRenderHook(obj, function(x) {
+#'   tagAppendAttributes(x, class = "extra")
+#' })
+#' spanExtra
+#'
+#' # Replace the previous render method
+#' # Should print a `div` with class `"extra"`
+#' divExtra <- tagRenderHook(obj, add = FALSE, function(x) {
+#'   tagAppendAttributes(x, class = "extra")
+#' })
+#' divExtra
+#'
+#' # Add more child tags
+#' spanExtended <- tagRenderHook(obj, function(x) {
+#'   tagAppendChildren(x, " ", tags$strong("bold text"))
+#' })
+#' spanExtended
+#'
+#' # Add a new html dependency
+#' newDep <- tagRenderHook(obj, function(x) {
+#'   fa <- htmlDependency(
+#'     "font-awesome", "4.5.0", c(href="shared/font-awesome"),
+#'     stylesheet = "css/font-awesome.min.css")
+#'   attachDependencies(x, fa, append = TRUE)
+#' })
+#' # Also add a jqueryui html dependency
+#' htmlDependencies(newDep) <- htmlDependency(
+#'   "jqueryui", "1.11.4", c(href="shared/jqueryui"),
+#'   script = "jquery-ui.min.js")
+#' # At render time, both dependencies will be found
+#' renderTags(newDep)$dependencies
+#'
+#' # Ignore the original tag and return something completely new.
+#' newObj <- tagRenderHook(obj, function(x) {
+#'   tags$p("Something else")
+#' })
+#' newObj
+tagRenderHook <- function(tag, func, add = TRUE) {
+  if (!is.function(func) || length(formals(func)) == 0) {
+    stop("`func` must be a function that accepts at least 1 argument")
+  }
+
+  prevFunc <- tag$.renderHook
+
+  if (!is.function(prevFunc) || !isTRUE(add)) {
+    tag$.renderHook <- func
+    return(tag)
+  }
+
+  tag$.renderHook <- function(x) {
```

Comment:
Instead of wrapping this stuff in closures, we should make `tag$.renderHook` a list, so that users are able to inspect render hook functions. (With closures, it's very difficult for a user to see how many render hooks there are, and what each one does.)

Then at render time, it can iterate through the list of functions, executing each one and passing along the returned value.

Reply:
similar to shiny route handlers?